packages/react-form-state/src/utilities.ts

```diff
@@ -27,3 +25,64 @@ export function replace<T>(array: T[], targetIndex: number, newValue: T) {
     return newValue;
   });
 }
+
+export function set<InputType extends Object>(
+  lhs: InputType,
+  path: string[],
+  value: any,
+) {
+  if (path.length === 0) {
+    return lhs;
+  } else if (path.length === 1) {
+    return {
+      ...(lhs as any),
+      [path[0]]: value,
+    };
+  } else {
+    const [current, ...rest] = path;
+    return {
+      ...(lhs as any),
+      [current]: set(lhs[current], rest, value),
+    } as InputType;
+  }
+}
+
+export function isEqual(value, baseline) {
+  if (typeof value !== typeof baseline) {
+    return false;
+  }
+  if (Array.isArray(value)) {
+    if (!Array.isArray(baseline)) {
+      return false;
+    }
+    if (value.length !== baseline.length) {
+      return false;
+    }
+    for (let iter = 0; iter < value.length; iter++) {
+      if (!isEqual(value[iter], baseline[iter])) {
+        return false;
+      }
+    }
+    return true;
+  }
+  if (typeof value === 'object') {
+    if (value === null) {
+      return baseline === null;
+    }
+    if (baseline === null) {
+      return false;
+    }
+    const keysInValue = Object.keys(value);
+    const keysInBaseline = Object.keys(baseline);
+    if (!isEqual(keysInValue, keysInBaseline)) {
+      return false;
+    }
+    for (const key of keysInValue) {
+      if (!isEqual(value[key], baseline[key])) {
+        return false;
+      }
+    }
+    return true;
+  }
+  return value === baseline;
```

Comment:
small thing, but we might want to put this check first to speed up prop comparisons where the object is reference equal across renders

Reply:
interesting. So we return `true` early if they are reference-equal and objects?

Reply:
well, reference equal at all I should think

Reply:
Ah, yes. But if they're objects and not reference equal, they may still be deep equal, whereas if they're not objects, then strict equality is the same as deep equality. I'll change this shortly.